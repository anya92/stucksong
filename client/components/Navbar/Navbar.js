import React, { Component } from 'react';
import { oneOfType, bool, object } from 'prop-types';
import { Link, NavLink } from 'react-router-dom';

import {
  Container,
  Brand,
  Bars,
  Links,
  LogoutLink,
} from '../../styles/navbar';

class Navbar extends Component {
  links = [
    { id: 0, link: 'top-tracks', name: 'tracks' },
    { id: 1, link: 'top-artists', name: 'artists' },
    { id: 2, link: 'recently-played', name: 'recently played' },
    { id: 3, link: 'create-playlist', name: 'create a playlist' },
  ]

  state = {
    open: false,
  }

  handleMenuClick = () => {
    this.setState(prevState => ({ open: !prevState.open }));
  }

  handleLinkClick = () => {
    this.setState(() => ({ open: false }));
  }

  render() {
    const { auth } = this.props;
    if (!auth) return <div />;
    return (
      <Container id="nav" open={this.state.open}>
        <Brand onClick={this.handleLinkClick}>
          <Link to="/">stuck<span>Song</span></Link>
        </Brand>
        <Bars className="mobile-menu" onClick={this.handleMenuClick}>
          <span />
        </Bars>
        <Links className="navbar-links">
          {
            this.links.map(({ id, link, name }) => (
              <NavLink
                key={id}
                to={`/${link}`}
                activeClassName="active"
                onClick={this.handleLinkClick}
              >{name}
              </NavLink>
            ))
          }
        </Links>
        <LogoutLink className="logout-link">
          <a href="/auth/logout">logout</a>
        </LogoutLink>
      </Container>
    );
  }
}

Navbar.propTypes = {
  auth: oneOfType([bool, object]).isRequired,
};

export default Navbar;
