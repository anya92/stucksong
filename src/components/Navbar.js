import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

class Navbar extends Component {
	constructor(props) {
		super(props);
	
		this.state = {
			path: this.props.location.pathname
		};
	}

	componentDidUpdate(prevProps, prevState) {
		if (this.props.location.pathname !== prevProps.location.pathname) {
			this.setState({
				path: this.props.location.pathname
			});
		}
	}

	render() {
		const { auth } = this.props;
		const { path } = this.state;
		return !auth ? <div></div> : (
			<div className="navbar-custom" ref={ref => (this.navbar = ref)}>
				<div className="navbar-custom__title">
					<Link to='/'>StuckSong</Link>
				</div>
				<div 
					className="navbar-custom__bars" 
					onClick={() => this.navbar.classList.toggle('open')}>
          <span className="navbar-custom__bars__bar"></span>
          <span className="navbar-custom__bars__bar"></span>
          <span className="navbar-custom__bars__bar"></span>
        </div>
								<div className="navbar-custom__user">
									<div className="navbar-custom__user__photo">
										{ auth.photo && <img src={auth.photo} alt={auth.username} /> }
									</div>
									<div className="navbar-custom__user__name">
										{auth.username}
									</div>
								</div>
								<div className="navbar-custom__links">
									<div className={`navbar-custom__links__link ${path === '/top-tracks' ? 'active' : ''}`} onClick={() => this.navbar.classList.remove('open')}>
										<Link to='/top-tracks'>Top Tracks</Link>
									</div>
									<div className={`navbar-custom__links__link ${path === '/top-artists' ? 'active' : ''}`} onClick={() => this.navbar.classList.remove('open')}>
										<Link to='/top-artists'>Top Artists</Link>
									</div>
									<div className={`navbar-custom__links__link ${path === '/recently-played' ? 'active' : ''}`} onClick={() => this.navbar.classList.remove('open')}>
										<Link to='/recently-played'>Recently Played</Link>
									</div>
									<div className={`navbar-custom__links__link ${path === '/create-playlist' ? 'active' : ''}`} onClick={() => this.navbar.classList.remove('open')}>
										<Link to='/create-playlist'>Create Playlist</Link>
									</div> <hr/>
									<div className='navbar-custom__links__link'>
										<a href="/auth/logout">Logout</a>
									</div>
								</div>

			</div>
		);
	}
}

function mapStateToProps(state) {
	return {
		auth: state.auth
	};
}


export default connect(mapStateToProps)(Navbar);
