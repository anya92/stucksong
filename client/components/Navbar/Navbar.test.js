import React from 'react';
import { shallow } from 'enzyme';
import { NavLink } from 'react-router-dom';

import Navbar from './Navbar';

describe('Navbar component', () => {
  it('renders Navbar without auth correctly', () => {
    const auth = false;
    const wrapper = shallow(<Navbar auth={auth} />);
    expect(wrapper).toMatchSnapshot();
  });

  let wrapper;
  let auth;
  beforeEach(() => {
    auth = {
      username: 'user1',
    };
    wrapper = shallow(<Navbar auth={auth} />);
  });

  it('renders Navbar with auth correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('opens and closes mobile navbar on mobile menu click', () => {
    expect(wrapper.state()).toEqual({ open: false });
    wrapper.find('.mobile-menu').simulate('click');
    expect(wrapper.state()).toEqual({ open: true });
    wrapper.find('.mobile-menu').simulate('click');
    expect(wrapper.state()).toEqual({ open: false });
  });

  it('closes mobile navbar on link click', () => {
    wrapper.find('.mobile-menu').simulate('click');
    expect(wrapper.state()).toEqual({ open: true });
    wrapper.find(NavLink).first().simulate('click');
    expect(wrapper.state()).toEqual({ open: false });
  });

  it('contains a logout link', () => {
    expect(wrapper.find('a').props().href).toBe('/auth/logout');
    expect(wrapper.find('a').text()).toBe('logout');
  });
});
