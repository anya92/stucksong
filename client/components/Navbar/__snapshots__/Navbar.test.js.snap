// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component renders with auth correctly 1`] = `
<glamorous(nav)
  id="nav"
  open={false}
>
  <glamorous(div)
    onClick={[Function]}
  >
    <Link
      replace={false}
      to="/"
    >
      stuck
      <span>
        Song
      </span>
    </Link>
  </glamorous(div)>
  <glamorous(div)
    className="mobile-menu"
    onClick={[Function]}
  >
    <span />
  </glamorous(div)>
  <glamorous(div)
    className="navbar-links"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      key="0"
      onClick={[Function]}
      to="/top-tracks"
    >
      tracks
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      key="1"
      onClick={[Function]}
      to="/top-artists"
    >
      artists
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      key="2"
      onClick={[Function]}
      to="/recently-played"
    >
      recently played
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      key="3"
      onClick={[Function]}
      to="/create-playlist"
    >
      create a playlist
    </NavLink>
  </glamorous(div)>
  <glamorous(glamorous(div))
    className="logout-link"
  >
    <a
      href="/auth/logout"
    >
      logout
    </a>
  </glamorous(glamorous(div))>
</glamorous(nav)>
`;

exports[`Navbar component renders without auth correctly 1`] = `<div />`;
